src/coreclr/tools/Common/TypeSystem/Interop/UnmanagedCallingConventions.cs

```diff
@@ -193,6 +194,53 @@ private static UnmanagedCallingConventions AccumulateCallingConventions(Unmanage
             return existing | addedCallConv.Value;
         }
 
+        public static EmbeddedSignatureData[] EncodeAsEmbeddedSignatureData(this UnmanagedCallingConventions callingConventions, TypeSystemContext context)
+        {
+            UnmanagedCallingConventions convention = (callingConventions & UnmanagedCallingConventions.CallingConventionMask);
+            UnmanagedCallingConventions modifiers = (callingConventions & UnmanagedCallingConventions.ModifiersMask);
+
+            UnmanagedCallingConventions platformDefault = GetPlatformDefaultUnmanagedCallingConvention(context);
+
+            int count = ((convention != platformDefault) ? 1 : 0) + BitOperations.PopCount((uint)modifiers);
+
+            if (count == 0)
+                return null;
+
+            EmbeddedSignatureData[] ret = new EmbeddedSignatureData[count];
+
+            int index = 0;
+
+            if (convention != platformDefault)
+            {
+                ret[index++] = CreateCallConvEmbeddedSignatureData(context, convention switch
+                {
+                    UnmanagedCallingConventions.Cdecl => "CallConvCdecl",
+                    UnmanagedCallingConventions.Stdcall => "CallConvStdcall",
+                    UnmanagedCallingConventions.Fastcall => "CallConvFastcall",
+                    UnmanagedCallingConventions.Thiscall => "CallConvThiscall",
+                    _ => throw new InvalidProgramException()
+                });
+            }
+
+            if ((modifiers & UnmanagedCallingConventions.IsMemberFunction) != 0)
+                ret[index++] = CreateCallConvEmbeddedSignatureData(context, "CallConvMemberFunction");
+
+            if ((modifiers & UnmanagedCallingConventions.IsSuppressGcTransition) != 0)
+                ret[index++] = CreateCallConvEmbeddedSignatureData(context, "CallConvSuppressGCTransition");
+
+            Debug.Assert(index == count);
+
+            return ret;
+
+            static EmbeddedSignatureData CreateCallConvEmbeddedSignatureData(TypeSystemContext context, string name)
+                => new()
+                {
+                    index = MethodSignature.IndexOfCustomModifiersOnReturnType,
+                    kind = EmbeddedSignatureDataKind.OptionalCustomModifier,
+                    type = context.SystemModule.GetKnownType("System.Runtime.CompilerServices", name)
```

Comment:
Nit: In theory we could have a corelib that doesn't define these but still uses the calling convention (just encoding them oldstyle). It's probably fine to have it fail and ask the user to add it to corelib.

Reply:
Yep